sopel/plugins/rules.py

```diff
@@ -1213,3 +1222,82 @@ def match_intent(self, intent):
         :rtype: bool
         """
         return bool(intent and self.INTENT_REGEX.match(intent))
+
+
+class FindRule(Rule):
+    """Anonymous find rule definition.
+
+    A find rule is like other anonymous rule with a twist: instead of maching
+    only once per IRC line, a find rule will execute for each non-overlapping
+    match for each of its regular expressions.
+
+    For example, to match for each word starting with the h letter in a line,
+    you can use the pattern ``h\\w+``:
+
+    .. code-block:: irc
+
+        <user> hello here
+        <Bot> Found the word "hello"
+        <Bot> Found the word "here"
+        <user> sopelunker, how are you?
+        <Bot> Found the word "how"
+
+    .. seealso::
+
+        This rule uses :func:`re.finditer`. To know more about how it works,
+        see the official Python documentation.
+
+    """
+    @classmethod
+    def from_callable(cls, settings, handler):
+        regexes = tuple(handler.find_rules)
+        kwargs = cls.kwargs_from_callable(handler)
+        kwargs['handler'] = handler
+
+        return cls(regexes, **kwargs)
+
+    def parse(self, text):
+        for regex in self._regexes:
+            for match in regex.finditer(text):
+                yield match
+
+
+class SearchRule(Rule):
+    """Anonymous search rule definition.
+
+    An anonymous search rule (or simply "a search rule") is like anonymous
+    rules with a twist: it will execute exactly once per regular expression
+    that match any part of a line, not just from the start.
```

Comment:
```suggestion
    A search rule is like an anonymous rule with a twist: it will execute
    exactly once per regular expression that matches anywhere in a line, not
    just from the start.
```